### ISSUE TITLE
`prefer-template` doesn't consider unary expressions in string concat as warning

### ISSUE THREAD

**Issue by Dashed** [eslint/eslint#3589]
`prefer-template` doesn't consider unary expressions in string concat as warning. 

For example:
```js
// EDIT: wrong one
//var foo = 'bar' + (typeof bar) + 'baz';

var foo = 'bar' + (typeof 100) + 'baz';
```

It can be rewritten as the following; but `prefer-template` doesn't emit a warning:

```js
var foo = `bar ${typeof 100} baz`;
```

/cc @mysticatea

**Comment by eslintbot** [comment]
Thanks for the issue! We get a lot of issues, so this message is automatically posted to each one to help you check that you've included all of the information we need to help you.

Reporting a bug? Please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
1. The source code that caused the problem
1. The configuration you're using (for the rule or your entire config file)
1. The actual ESLint output complete with line numbers

Requesting a new rule? Please be sure to include:

1. The use case for the rule - what is it trying to prevent or flag?
1. Whether the rule is trying to prevent an error or is purely stylistic
1. Why you believe this rule is generic enough to be included

Requesting a feature? Please be sure to include:

1. The problem you want to solve (don't mention the solution)
1. Your take on the correct solution to problem

Including this information in your issue helps us to triage it and get you a response as quickly as possible.

Thanks!

**Comment by mysticatea** [comment]
Oh, `var foo = 'bar' + (typeof bar) + 'baz';` is warned correctly.

```js
/*eslint prefer-template:2, no-useless-concat:2*/
var foo = 'bar' + 100 + 'baz';
var foo = 'bar' + (typeof 100) + 'baz';
var foo = 'bar' + (typeof bar) + 'baz';
```

```
> eslint -v
v1.3.1
> eslint test.js --no-eslintrc

test.js
  2:17  error  Unexpected string concatenation of literals  no-useless-concat
  2:23  error  Unexpected string concatenation of literals  no-useless-concat
  4:11  error  Unexpected string concatenation              prefer-template

✖ 3 problems (3 errors, 0 warnings)
```

Let's consider about literal of non string.

Should `prefer-template` warn the following pattern?

```js
var foo = 'bar' + 100 + 'baz';
var foo = 'bar' + (typeof 100) + 'baz';
```

If we make `no-useless-concat` ignoring literal of non string in #3575, I thought `prefer-template` should handle those.

**Comment by michaelficarra** [comment]
This rule should simply report all uses of `+` with a literal string (or untagged template?) operand.

**Comment by Dashed** [comment]
sorry. copied the wrong example. I meant `var foo = 'bar' + (typeof 100) + 'baz';`.

IMO, these should be warnings since non-literals are being concatenated:
```js
var foo = 'bar' + 100 + 'baz';
var foo = 'bar' + (typeof 100) + 'baz';
```

**Comment by mysticatea** [comment]
Thank you.

If `no-useless-concat` is kept the current behavior, how do you think?

```js
var foo = 'bar' + 100 + 'baz';          // suggest 'bar100baz' by no-useless-concat
var foo = 'bar' + (typeof 100) + 'baz'; // suggest 'barnumberbaz' by no-useless-concat
var foo = 'bar' + (typeof bar) + 'baz'; // suggest `bar${typeof bar}baz` by prefer-template
```

**Comment by Dashed** [comment]
For `var foo = 'bar' + (typeof 100) + 'baz';`, this shouldn't trigger `no-useless-concat`; but it triggers `prefer-template`.

**Comment by nzakas** [comment]
What are we saying here? Is there a bug? Is this requesting an enhancement? @mysticatea

**Comment by Dashed** [comment]
@nzakas This is a bug for `prefer-template`.

**Comment by nzakas** [comment]
@Dashed can you explain more? I'm having a hard time following this thread. What is the bug? (Note the second comment that asked for more information.) Before making any change, I'd like to see a clear description of the bug.

**Comment by Dashed** [comment]
The bug is that `prefer-template` should consider the following as a warning:

```js
var foo = 'bar' + (typeof 100) + 'baz';
```

since it can be rewritten as

```js
var foo = `bar ${typeof 100} baz`;
```

This is a bug stemming from https://github.com/eslint/eslint/pull/3513; and the PR made `prefer-template` not consider unary expressions as string concatenation as something that can be rewritten into template strings.

Hope that clarifies it.

----

I opened this issue at the request of @mysticatea: https://github.com/eslint/eslint/pull/3513#issuecomment-136073085

**Comment by Dashed** [comment]
I amended the first post to clarify the example I posted.

**Comment by nzakas** [comment]
Thanks, that helps a lot! So @mysticatea, you're fixing this to so `prefer-template` will flag that expression?

**Comment by Dashed** [comment]
@mysticatea  has a PR (https://github.com/eslint/eslint/pull/3602) that polishes the semantics for `prefer-template` and `no-useless-concat`

**Comment by mysticatea** [comment]
I'm sorry for my late response.

I'm not sure this is a bug or an enhancement. 
This purpose is to fill in a hole which is made in #3575.

```js
var foo = 'bar' + 'baz';       // is reported by no-useless-concat
var foo = 'bar' + 100 + 'baz'; // is reported by no-useless-concat
var foo = 'bar' + abc + 'baz'; // is reported by prefer-template
```

Above is the original behavior.
But as the result of #3575, those become below:

```js
var foo = 'bar' + 'baz';       // is reported by no-useless-concat
var foo = 'bar' + 100 + 'baz'; // 
var foo = 'bar' + abc + 'baz'; // is reported by prefer-template
```

Then `prefer-template` is extended concatenation with numbers in this issue.

```js
var foo = 'bar' + 'baz';       // is reported by no-useless-concat
var foo = 'bar' + 100 + 'baz'; // is reported by prefer-template
var foo = 'bar' + abc + 'baz'; // is reported by prefer-template
```

**Comment by nzakas** [comment]
Okay, so we should consider these two separate changes:

1. There was a bug in `no-useless-concat` where it would flag non-string values.
1. There was a bug in `prefer-template` where it was not flagging literals.

We need to be careful about making changes to existing rules without being certain why we're doing it, as these could be breaking changes to existing users unless we identified them as bugs.

**Comment by Dashed** [comment]
The rule `no-useless-concat` is just being redefined from its original implementation; is this still considered a bug?

**Comment by nzakas** [comment]
It's not being redefined. The rule's purpose was to flag unnecessary concatenation and it was incorrectly flagging necessary concatenation.
